### PR TITLE
Harmonize behavior scanning expired QR (EXPOUREAPP-2700)

### PR DESCRIPTION
**Pull request by AndroidMedaGalaxy**
### Description
If Expired QR code is scanned, a message is displayed only once where user is redirected to the previous screen and they can try scanning the QR again.

### Screenshot
QR Expired harmonization 

### Steps to reproduce
Step 1: Scan an expired QR code
Step 2: Observe that an error dialog is displayed
Step 3: QR code can be scanned again or the user can decide to come back to the Home screen

### CODE REVIEW

**Review by janetback — DISMISSED** [review]
UA approved

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/fragment/SubmissionQRCodeScanFragment.kt [hunk 1]
```diff
@@ -123,10 +126,21 @@ class SubmissionQRCodeScanFragment : Fragment(R.layout.fragment_submission_qr_co
...
+                        if (uiState == DeviceUIState.PAIRED_REDEEMED) {
+                            showRedeemedTokenWarningDialog()
+                            submissionViewModel.deregisterTestFromDevice()
+                            goBack()
+                        } else {
+                            doNavigate(
+                                SubmissionQRCodeScanFragmentDirections
+                                    .actionSubmissionQRCodeScanFragmentToSubmissionResultFragment()
+                            )
+                        }
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/fragment/SubmissionQRCodeScanFragment.kt:141]: in my opinion this logic should be moved to the view model, then send events for navigation to fragment via life data

**Review by chris-cwa — DISMISSED** [review]

**Comment by chris-cwa** [comment]
loving the gif!

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/fragment/SubmissionQRCodeScanFragment.kt (hunk 1 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/fragment/SubmissionQRCodeScanFragment.kt]: Hi Chris,

the logic has been revamped. Please take a look again. and let me know if you have any question.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanFragment.kt [hunk 2]
```diff
@@ -59,21 +60,29 @@ class SubmissionQRCodeScanFragment : Fragment(R.layout.fragment_submission_qr_co
         binding.submissionQrCodeScanViewfinderView.setCameraPreview(binding.submissionQrCodeScanPreview)
 
         viewModel.scanStatus.observeEvent(viewLifecycleOwner) {
```
Comment (kolyaopahle) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanFragment.kt:62]: as this is only observed here and the value does not matter we could also replace scan status with a `SingleLiveEvent<Unit>`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt [hunk 3]
```diff
@@ -6,29 +6,39 @@ import com.squareup.inject.assisted.AssistedInject
...
+class SubmissionQRCodeScanViewModel @AssistedInject constructor() :
+    CWAViewModel() {
     val routeToScreen: SingleLiveEvent<SubmissionNavigationEvents> = SingleLiveEvent()
     private val _scanStatus = MutableLiveData(Event(ScanStatus.STARTED))
 
     val scanStatus: LiveData<Event<ScanStatus>> = _scanStatus
+    val showRedeemedTokenWarning = SingleLiveEvent<Unit>()
+
+    open class InvalidQRCodeException : ReportedIOException(
+        ErrorCodes.CWA_WEB_REQUEST_PROBLEM.code, "error in qr code"
```
Comment (kolyaopahle) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt:34]: I don't think this is a WEB_REQUEST_PROBLEM, you could declare it to simply extend Exception as it is only used internally

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt [hunk 4]
```diff
@@ -19,41 +18,31 @@ object SubmissionService {
...
         val (registrationToken, testResult) =
             playbook.initialRegistration(
                 guid,
                 VerificationKeyType.GUID
             )
-
         LocalData.registrationToken(registrationToken)
         deleteTestGUID()
         SubmissionRepository.updateTestResult(testResult)
+        LocalData.devicePairingSuccessfulTimestamp(System.currentTimeMillis())
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt:30]: we try to use timestamped now everywhere

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt [hunk 5]
```diff
@@ -19,41 +18,31 @@ object SubmissionService {
...
         val (registrationToken, testResult) =
             playbook.initialRegistration(
                 tan,
                 VerificationKeyType.TELETAN
             )
-
         LocalData.registrationToken(registrationToken)
         deleteTeleTAN()
         SubmissionRepository.updateTestResult(testResult)
+        LocalData.devicePairingSuccessfulTimestamp(System.currentTimeMillis())
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt:44]: we try to use timestamped now everywhere

**Review by kolyaopahle — CHANGES_REQUESTED** [review]
Two small nitpicks otherwise LGTM 👍

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt [hunk 6]
```diff
@@ -37,10 +47,10 @@ class SubmissionQRCodeScanViewModel @AssistedInject constructor() : CWAViewModel
     val registrationState = MutableLiveData(ApiRequestState.IDLE)
     val registrationError = SingleLiveEvent<CwaWebException>()
 
-    fun doDeviceRegistration() = launch {
+    private fun doDeviceRegistration(scanResult: QRScanResult) = launch {
         try {
             registrationState.postValue(ApiRequestState.STARTED)
-            SubmissionService.asyncRegisterDevice()
+            checkTestResult(SubmissionService.asyncRegisterDeviceViaGUID(scanResult.guid!!))
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt:53]: I would prefer an explicit null-check here

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanFragment.kt [hunk 7]
```diff
@@ -59,21 +60,29 @@ class SubmissionQRCodeScanFragment : Fragment(R.layout.fragment_submission_qr_co
...
+        viewModel.showRedeemedTokenWarning.observe2(this) {
+            val dialog = DialogHelper.DialogInstance(
+                requireActivity(),
+                R.string.submission_error_dialog_web_tan_redeemed_title,
+                R.string.submission_error_dialog_web_tan_redeemed_body,
+                R.string.submission_error_dialog_web_tan_redeemed_button_positive
+            )
+
+            DialogHelper.showDialog(dialog)
+            viewModel.deregisterTestFromDevice()
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanFragment.kt:77]: this should be done in ViewModel, when the value of the life data, that is being observed here, is posted

**Review by chris-cwa — CHANGES_REQUESTED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt (hunk 6 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt:53]: maybe this calls for a refactoring of `QRScanResult` as `guid` cannot be null if `isValid` is `true` as seen [here](https://github.com/corona-warn-app/cwa-app-android/blob/fix/harmonize-behaviour-scanning-expired-QR-EXPOSUREAPP-2700/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/QRScanResult.kt#L7-L8)

**Comment by AlexanderAlferov** [comment]
Tested - looks good.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt (hunk 3 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt:34]: @kolyaopahle  How do you feel about a generic Exception here?

  open class InvalidQRCodeException : Exception("error in qr code" )

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt (hunk 4 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt:30]: done

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt (hunk 5 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/service/submission/SubmissionService.kt:44]: done

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanFragment.kt (hunk 7 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanFragment.kt]: That's been done. Perfect point of moving it to viewmodel.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanFragment.kt (hunk 2 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanFragment.kt]: this has been implmented. please check it out.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt (hunk 3 above)
Comment (ritsam) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/scan/SubmissionQRCodeScanViewModel.kt]: Used the about mentioned implmentation. check it out when you can.

**Review by AlexanderAlferov — APPROVED** [review]
Retested. LGTM!

**Review by kolyaopahle — APPROVED** [review]

**Review by chris-cwa — APPROVED** [review]

**Pull request merged**